### ISSUE TITLE
Paths containing characters other than 7-bit ASCII cause failure

### ISSUE THREAD

**Issue by DavidHollman** [NREL/EnergyPlus#10487]
Issue overview
--------------
Running EnergyPlus with a command like the following will cause it to fail:

`energyplus.exe -d D:\ß D:\ß\sample.idf`

Output:
```
EnergyPlus Starting
EnergyPlus, Version 23.2.0-7636e6b3e9, YMD=2024.05.01 11:29
**FATAL:The SQLite database failed to open.
EnergyPlus Run Time=00hr 00min  0.19sec
Program terminated: EnergyPlus Terminated--Error(s) Detected.
```

If however the folder were named `B` instead of `ß`, all else being equal, it would work normally.

The folder `D:\ß` actually has to exist for this error to occur. The contents of the IDF file don't seem to matter.

---

Notes - I think the problem occurs when using a folder path having any ASCII/Unicode character [over 7F](https://www.ascii-code.com/). Furthermore, varied higher bitness characters fail in different ways.

I tried the following examples as folder names (all char values in hex):

 - `B` (ASCII 42) - runs normally
 - `` (ASCII 7F or "del" - the last 7-bit value) - not a printable character, but still runs normally
 - `€` (ASCII 80 - the first 8-bit value) - fails with SQLite error
 - `Â` ([Unicode 00C2](https://unicode.scarfboy.com/?s=%C3%82)) - fails with SQLite error
 - `ß` ([Unicode 00DF](https://unicode.scarfboy.com/?s=%C3%9F)) - fails with SQLite error (the example given above)
 - `Б` ([Unicode 0411](https://unicode.scarfboy.com/?s=%D0%91)) - fails but in a different way - see below

For the example of `Б` which is 2 bytes in Unicode, EnergyPlus terminates without ever writing anything to the terminal, unlike the 128-bit ASCII and 1-byte Unicode examples which show the SQLite error.

---

**Note on relevance --** This was discovered because I support users who can have usernames containing Unicode characters, and therefore their local paths also contain that character. (_Just to point out that this is not at all a hypothetical problem._)

---

We used to use EnergyPlus 9.5 before updating to 23.2, and to my knowledge never ran into this before. Perhaps this is related to some of the command-line updates for CLI11 in 23.2 (mentioned in https://github.com/NREL/EnergyPlus/issues/10322) although that's just a guess and might not actually make sense given the SQLite error.  _[EDIT - maybe similar issues did exist earlier, I'll add a follow up post]_

### Details
Some additional details for this issue (if relevant):
 - Platform: Windows 10, 11
 - Version of EnergyPlus: 23.2

**Comment by DavidHollman** [comment]
A colleague ran tests with earlier versions and found similar problems, though the exact error messages are not all the same.

# 9.5

```
C:\>EnergyPlus\EnergyPlusV9-5-0\EnergyPlus -d d:\ß\  d:\ß\Test_V950.idf
EnergyPlus Starting
EnergyPlus, Version 9.5.0-de239b2e5f, YMD=2024.05.01 11:56
  File "<string>", line 1
SyntaxError: (unicode error) 'utf-8' codec can't decode byte 0xdf in position 0: unexpected end of data
**FATAL:ERROR adding "d:\\▀" to the sys.path in Python
EnergyPlus Run Time=00hr 00min  0.20sec
Program terminated: EnergyPlus Terminated--Error(s) Detected.
```

# 9.6
```
C:\>EnergyPlus\EnergyPlusV9-6-0\EnergyPlus -d d:\ß\  d:\ß\Test_V960.idf
EnergyPlus Starting
EnergyPlus, Version 9.6.0-f420c06a69, YMD=2024.05.01 11:56
  File "<string>", line 1
    sys.path.insert(0, "d:\\�")
                            ^
SyntaxError: (unicode error) 'utf-8' codec can't decode byte 0xdf in position 0: unexpected end of data
**FATAL:ERROR adding "d:\\ß" to the sys.path in Python
EnergyPlus Run Time=00hr 00min  0.09sec
Program terminated: EnergyPlus Terminated--Error(s) Detected.
```
# 22.1
```
C:\>EnergyPlus\EnergyPlusV22-1-0\EnergyPlus -d d:\ß\  d:\ß\Test_V2210.idf
EnergyPlus Starting
EnergyPlus, Version 22.1.0-ed759b17ee, YMD=2024.05.01 11:57
  File "<string>", line 1
SyntaxError: (unicode error) 'utf-8' codec can't decode byte 0xdf in position 0: unexpected end of data
**FATAL:ERROR adding "d:\\ß" to the sys.path in Python
EnergyPlus Run Time=00hr 00min  0.09sec
Program terminated: EnergyPlus Terminated--Error(s) Detected.
```

# 22.2
```
C:\>EnergyPlus\EnergyPlusV22-2-0\EnergyPlus -d d:\ß\  d:\ß\Test_V2220.idf
EnergyPlus Starting
EnergyPlus, Version 22.2.0-c249759bad, YMD=2024.05.01 11:57
  File "<string>", line 1
SyntaxError: (unicode error) 'utf-8' codec can't decode byte 0xdf in position 0: unexpected end of data
**FATAL:ERROR adding "d:\\ß" to the sys.path in Python
EnergyPlus Run Time=00hr 00min  0.09sec
Program terminated: EnergyPlus Terminated--Error(s) Detected.
```

**Comment by jmarrec** [comment]
I cannot reproduce this (tried 24.1.0 and 23.1.0)

```
cd Downloads\test
mkdir ß
cp C:\EnergyPlusV24-1-0\ExampleFiles\1ZoneUncontrolled.idf ß\in.idf
```

```
C:\EnergyPlusV24-1-0\energyplus -D -d C:\Users\julien\Downloads\test\ß C:\Users\julien\Downloads\test\ß\in.idf 
[...]
EnergyPlus Completed Successfully
```

tested in pwsh and cmd.exe.

**Comment by jmarrec** [comment]
cd to C:/

```
mkdir ß
cp C:\EnergyPlusV24-1-0\ExampleFiles\1ZoneUncontrolled.idf ß\in.idf
C:\EnergyPlusV24-1-0\energyplus -D -d C:\ß C:\ß\in.idf
[...]
EnergyPlus Completed Successfully
```

**Comment by jmarrec** [comment]
> The folder `D:\ß` actually has to exist for this error to occur

Given that sample.idf is in that `ß` folder, I don't see how you could not have it exist?

**Comment by jmarrec** [comment]
Hum, could you do `chcp` in a terminal?

```
> chcp
Active code page: 437
```

Even if I do chcp 65001(or 850 1252) then retry, it works though.

**Comment by jmarrec** [comment]
> `ß` ([Unicode 00DF](https://unicode.scarfboy.com/?s=%C3%9F)) - fails with SQLite error (the example given above)

I missed that technicality, since the message said "The contents of the IDF file don't seem to matter". I guess 1ZoneUncontrolled.idf does not have Output:SQLite, so that's why I wasn't seeing this.

(a .zip with the test case would have helped reproduce)

**Comment by jmarrec** [comment]
I think I can reasonably make the 1-bye unicode support work (eg: ß), but not the Б ([Unicode 0411](https://unicode.scarfboy.com/?s=%D0%91)). That one would be way too much hassle. 

That one requires too much branching for win32, including the calls to fopen needed to be replaced by _wfopen or it'll throw No mapping for the Unicode character exists in the target multi-byte code page, and changing all calls to fs::path::string to be  native / wstring one or the wide to narrow conversion fails.

**Comment by DavidHollman** [comment]
Hi @jmarrec thanks for looking into this one!  

---

When I wrote that the IDF contents "didn't matter" what I was getting at was that I was seeing the pattern of problems with any IDF file I had tried. 

But all the samples I used may still have had some similarities which I wasn't looking for. I certainly should have attached the sample I used! I no longer have it.

---

By "_The folder D:\ß actually has to exist_" -- it's possible to run EnergyPlus with arguments to a file location that does not really exist. I initially wondered if that would be enough to trigger this problem, which it was not. But it still seemed like potentially useful information to narrow down the cause. 

---

`chcp` gives me:

```
C:\Users\usernameßabcd>chcp
Active code page: 437
```

On both systems where I tried this. I think that makes sense, the user account didn't have any special region/language settings.

The other code pages you tried - how did you know which ones to use?  Can you confirm that avoided the error?

---

Although anything that reduces the likelihood of a problem will be helpful.  But if it can't be done for all possible valid characters then we may have to look for some kind of workaround anyway. 

We could try to put the file in some location which only consists of ASCII characters. But I don't actually know if such a guarantee is possible on Windows. (Maybe changing the code page first could be a way around this.)

I'm surprised this hasn't come up before, I would have thought this circumstance would be common in many countries... I wonder why not?

**Comment by jmarrec** [comment]
As someone who is native to a country that has accents, I learned a long time ago that sticking to ASCII only is the best option, especially when it comes to picking a **username** - since that defines your `C:\Users\username` folder. Even official Microsoft programs would sometimes fail miserably if you didn't (and changing your username folder after the fact is a massive pain, it's faster to just create a new profile and just delete the old one).

(I also follow the convention that, in addition to non ASCII characters, I never put spaces in any folder or file names, it's a common source of problems as well.
I also try to use short folder structures, to limit the possibility of reaching the 260 character limits in windows... by default the LONG PATHS option is disabled even on windows 10)

I'm not saying this is acceptable, just sharing the reality of things.

And this problem has been reported numerous times on unmethours.com (sometimes for EnergyPlus, sometimes for OpenStudio). One example: https://unmethours.com/question/92675/trying-to-run-a-simulation-through-the-command-line-on-openstudio/ , another: https://unmethours.com/question/21147/energyplus-error-message/

**Comment by jmarrec** [comment]
@DaveInCaz Could you run some tests with https://github.com/jmarrec/EnergyPlus/releases/tag/v24.2.0-windows_non_ascii please? I think I fixed it, including the 2 bytes unicode such as Б ([Unicode 0411](https://unicode.scarfboy.com/?s=%D0%91)

**Comment by DavidHollman** [comment]
@jmarrec thanks for working on this.

I just tried to reproduce the problem with the 24.2.0 release as you suggested and it worked great. **I no longer can detect any problem.**

I first repeated my original tests with 23.2 just to make sure I was running them correctly, and then did the same with 24.2. _To confirm, the failing commands with 23.2 would not fail with 24.2_.

Sample IDF file I had used is now attached: [sample.zip](https://github.com/user-attachments/files/16234890/sample.zip)


---

BTW I personally agree with your sentiment about sensibly avoiding problems due to "special" characters, spaces, etc. But in the case of the software I work on where EnergyPlus is incorporated as an API, our end users would probably not be aware of this; or even have minimal control over their PC which typically would have been setup for them by IT, etc.

Thanks again!
